lib/kitchen/transport/exec.rb

```diff
@@ -40,19 +40,105 @@ class Connection < Kitchen::Transport::Base::Connection
         def execute(command)
           return if command.nil?
 
-          run_command(command)
+          if host_os_windows?
+            run_command(run_from_file_command(command))
+            close
+          else
+            run_command(command)
+          end
+        end
+
+        def close
+          if host_os_windows?
+            FileUtils.remove(exec_script_file)
+          end
         end
 
         # "Upload" the files by copying them locally.
         #
         # @see Base#upload
         def upload(locals, remote)
-          FileUtils.mkdir_p(remote)
+          # evaluate $env:temp on Windows
+          real_remote = remote.to_s == "\$env:TEMP\\kitchen" ? kitchen_temp : remote
+          FileUtils.mkdir_p(real_remote)
           Array(locals).each do |local|
-            FileUtils.cp_r(local, remote)
+            FileUtils.cp_r(local, real_remote)
+          end
+        end
+
+        # (see Base#init_options)
+        def init_options(options)
+          super
+          @instance_name = @options.delete(:instance_name)
+          @kitchen_root = @options.delete(:kitchen_root)
+        end
+
+        private
+
+        # @return [String] display name for the associated instance
+        # @api private
+        attr_reader :instance_name
+
+        # @return [String] local path to the root of the project
+        # @api private
+        attr_reader :kitchen_root
+
+        # Takes a long command and saves it to a file and uploads it to
+        # the test instance. Windows has cli character limits.
+        #
+        # @param command [String] a long command to be saved and uploaded
+        # @return [String] a command that executes the uploaded script
+        # @api private
+        def run_from_file_command(command)
+          if logger.debug?
+            debug("Creating exec script for #{instance_name} (#{exec_script_file})")
+            debug("Executing #{exec_script_file}")
+          end
+          File.open(exec_script_file, "wb") { |file| file.write(command) }
+          %{powershell -file "#{exec_script_file}"}
+        end
+
+        # @return [String] evaluated $env:temp variable
+        # @api private
+        def kitchen_temp
+          "#{ENV["temp"]}/kitchen"
+        end
+
+        # @return [String] name of script using instance name
+        # @api private
+        def exec_script_name
+          "#{instance_name}-exec-script.ps1"
+        end
+
+        # @return [String] file path for exec script to be run
+        # @api private
+        def exec_script_file
+          File.join(kitchen_root, ".kitchen", exec_script_name)
+        end
+
+        def host_os_windows?
```

Comment:
FWIW this was copied from the winrm transport file. It might be better if I moved this to a more central location so it could be used globally if needed.